### ISSUE TITLE
make the 'insert' and 'update' methods not modify the past Map

### ISSUE THREAD

**Issue by insinfo** [mongo-dart/mongo_dart#236]
make the 'insertOne' and 'replaceOne' ... methods not modify the past Map, 
to prevent the error, "MongoError: The (immutable) field '_id' was found to have been altered"

```dart
  final cupom = Cupom.fromMap(req.bodyAsMap);
  cupom.setInfoByLang('pt');
  final repository = CupomRepository();
  cupom.quantidade = cupom.quantidade - 1;
  final map = cupom.toMap();
  var result = await repository.createCupomUsado(map);
  //to prevent the error, "MongoError: The (immutable) field '_id' was found to have been altered"
  map.remove('_id');    
  result = await repository.updateById(map, cupom.id);

```

**Comment by giorgiofran** [comment]
Yes, `insertOne` adds an `_id` field **_if not specified_**. Why should it give you a problem?

**Comment by insinfo** [comment]
this caused me a problem, because I do an insertion and then an update of the same map in different collections, the first collection being a history of changes, and the second collection is the main one, but when I go to save in the second collection the error is due to that the first insert modifies the Map by putting an _id

**Comment by insinfo** [comment]
maybe a solution for this would be the insertOne method to clone the Map so that it doesn't modify the original

**Comment by giorgiofran** [comment]
Let's say that I have a map (`myMap`):` <String, dynamic>{'name': 'Isaque', 'familyName':'Neves}`
Now I do the first insert in a collection, let's say `history`.
As MongoDb **_requires_** an `_id` field we have two ways of set the '_id':
1) before insertion create an `ObjectId` (or any key you like) and assign it to the map:
 ```dart
var documentKey = ObjectId();
myMap['_id'] = documentKey;
await history.insertOne(myMap);
```
This way I have the map ready to be inserted in another collection (`actual`)
```dart
await actual.insertOne(myMap);
```
If you need to update a document on `actual` simply use the map you have
```dart
actual.replaceOne( where.id(myMap['_id']),  myMap);
```
or 
```dart
actual.replaceOne( where.id(documentKey),  myMap);
```
2) I you do not want to insert manually the `_id` field, `insertOne `will do this for you.
The result is that `myMap` will have the` _id` set and you will be able to use it for the update.

So, filling the `_id` field in `insertOne` is intentional and useful.

**Comment by giorgiofran** [comment]
Sorry, if I cannot help you anymore, but it is still unclear to me what you would like to do.
If you could post an example without using your framework it could be useful.

**Comment by insinfo** [comment]
Let us assume this situation:

### I have a Map so it's also stored in the 'coupons' collection :
```dart
var data = {"id":123456,"name":"Isaque", "age":36}
```
### I want to be included in the 'usedCoupons' collection
```dart
  await db.collection('usedCoupons').insertOne(data);
```
### and then I want to update the 'coupons' collection

```dart
await db.collection( 'coupons').replaceOne({'id':123456}, date);
```

### This code will fail as the insertOne method added an '_id' to Map data
### A possible solution would be the insertOne method to clone the variable data like this
```dart
final Map<String, dynamic> documentClone = {...document};
```

**Comment by insinfo** [comment]
I don't think you understand, I don't use the "_id" field, because I have my own method of generating unique "id", so I use an "id" field instead of "_id", my solution was to make a wrapper on top of the driver so that the insertion into the database is done without changing the Map passed via parameter, this was done easily through the spreed operator cloning the Map, by design I think the driver should not modify the instance of Map passed by parameter, it should clone the Map and modify the clone and not the original instance. my problem is not mongo putting the "_id" field, the problem is the driver modifying the Map instance passed via parameter

**Comment by giorgiofran** [comment]
What I do not understand is why you do not use the `_id` field if you already have an unique identifier.
The `_id` field can be of any type you like, not necessarily an `ObjectId`.
